### PR TITLE
Created dry run flag for collector

### PR DESCRIPTION
**Pull request by Chinwendu20**
Signed-off-by: Maureen <amaka013@gmail.com>

**Description:** Created dry run flag for the collector

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/4613

**Documentation:** Updated readme of service package but would create an issue on opentelemetry.io to update documentation

### CODE REVIEW

**Comment by Chinwendu20** [comment]
Heyy, just discovered there were some recent changes to the main branch that would affect this PR. I would make some changes and push

File: service/config_provider.go [hunk 1]
```diff
@@ -60,6 +60,8 @@ type ConfigProvider interface {
 	//
 	// Should never be called concurrently with itself or Get.
 	Shutdown(ctx context.Context) error
+	// DryRunGet validates the configuration, prints out all related errors without running the collector
+	DryRunGet(ctx context.Context, factories component.Factories) (*Config, error, bool)
```
Comment (bogdandrutu) [inline service/config_provider.go:64]: Should "dry run" be a flag or a command? 

/cc @codeboten @jpkrohling who also reviewed https://github.com/open-telemetry/opentelemetry-collector/pull/6322

Comment (codeboten) [reply service/config_provider.go:64]: i would expect it to be a flag

File: config/moved_config.go [hunk 2]
```diff
@@ -153,6 +153,97 @@ func (cfg *Config) validateService() error {
 	}
 	return nil
 }
+func (cfg *Config) DryRunValidate() {
+	// Currently, there is no default receiver enabled.
+	// The configuration must specify at least one receiver to be valid.
+	if len(cfg.Receivers) == 0 {
```
Comment (jpkrohling) [inline config/moved_config.go:159]: Each one of the conditions needs its test to ensure its correctness.

File: service/internal/configunmarshaler/dry_run_unmarshaller.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+var ExtensionValidateErrors []ExtensionValidateError
+var ExportersValidateErrors []ExporterValidateError
+var ReceiversValidateErrors []ReceiverValidateError
+var ProcessorsValidateErrors []ProcessorValidateError
+
+func (ConfigUnmarshaler) DryRunUnmarshal(v *confmap.Conf, factories component.Factories) (*config.Config, error, bool) {
+	errorcheck := false
+	if err := v.Unmarshal(&rawCfg, confmap.WithErrorUnused()); err != nil {
```
Comment (jpkrohling) [inline service/internal/configunmarshaler/dry_run_unmarshaller.go:43]: Using a var that is defined in an unrelated file would be confusing. If you have to use the same var, it might make sense to split this into its own source file.

File: service/internal/configunmarshaler/dry_run_unmarshaller.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+import (
+	"fmt"
+	"reflect"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/confmap"
+)
+
+type ExtensionValidateError struct {
```
Comment (jpkrohling) [inline service/internal/configunmarshaler/dry_run_unmarshaller.go:12]: If those four types have the same definition, perhaps you can have only one? `ComponentValidationError` would probably be a good candidate here.

File: service/internal/configunmarshaler/dry_run_unmarshaller.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	Err       string
+}
+
+type ProcessorValidateError struct {
+	Component string
+	Id        config.ComponentID
+	Err       string
+}
+
+var ExtensionValidateErrors []ExtensionValidateError
```
Comment (jpkrohling) [inline service/internal/configunmarshaler/dry_run_unmarshaller.go:36]: You can still have four vars, each being a slice of `ComponentValidationError`s

**Comment by Chinwendu20** [comment]
Still have to make some changes as recent changes in the main branch has affected my most recent push as well. I will try to finish this one up as fast as possible.

**Comment by Chinwendu20** [comment]
Had  a working solution earlier today just to find out that there are new changes to the the main branch later today hence this conflict.

**Comment by Chinwendu20** [comment]
So the problem is that I am creating a new function for validate that just prints validation errors when it is encountered as opposed to returning quickly so if there is a change in the implementation of the main validate function I would have to change mine. I do not know if there is a smarter way of going about this

**Comment by jpkrohling** [comment]
> So the problem is that I am creating a new function for validate that just prints validation errors when it is encountered as opposed to returning quickly

It's okay -- I prefer to have all errors listed instead of failing fast in this case.

> if there is a change in the implementation of the main validate function I would have to change mine

Not quite sure what you meant here.

**Comment by Chinwendu20** [comment]
> > So the problem is that I am creating a new function for validate that just prints validation errors when it is encountered as opposed to returning quickly
> 
> It's okay -- I prefer to have all errors listed instead of failing fast in this case.
> 
> > if there is a change in the implementation of the main validate function I would have to change mine
> 
> Not quite sure what you meant here.

I am giving a reason why I have had to keep changing my code and why some checks are failing It is because at the time there were changes to the validate function after submitting my PR . I would go at it again
